### PR TITLE
Forbid paths in timezone names

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Forbid paths in timezone names. OSS-Fuzz found that it can lead to arbitrary file open (unfortunately link to oss-fuzz was expired)

### CODE REVIEW

File: src/DataTypes/TimezoneMixin.h [hunk 1]
```diff
@@ -29,4 +37,17 @@ class TimezoneMixin
...
     const DateLUTImpl & time_zone;
     const DateLUTImpl & utc_time_zone;
+
+private:
+    static const String & checkTimezoneName(const String & timezone_name)
+    {
+        const char * forbidden_patterns[] = {"/", "../", "./", "~/"};
+        for (const auto & pattern : forbidden_patterns)
+        {
+            if (timezone_name.starts_with(pattern))
```
Comment (tavplubix) [inline src/DataTypes/TimezoneMixin.h:47]: Can it contain a forbidden pattern in the middle?

Comment (Avogar) [reply src/DataTypes/TimezoneMixin.h:47]: Good catch, we should try to normalize the path before this check. I will do it

Comment (Avogar) [reply src/DataTypes/TimezoneMixin.h]: I didn't find a simple way to normalize a path. I think checking `"../"` substring will be enough.

File: src/DataTypes/TimezoneMixin.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+{
+    const char * forbidden_beginnings[] = {"/", "./", "~"};
+    for (const auto & pattern : forbidden_beginnings)
+    {
+        if (timezone_name.starts_with(pattern))
+            throw DB::Exception(DB::ErrorCodes::BAD_ARGUMENTS, "Timezone name cannot start with '{}'", pattern);
+    }
+
+    if (timezone_name.find("../") != std::string::npos)
+        throw DB::Exception(DB::ErrorCodes::BAD_ARGUMENTS, "Timezone name cannot contain pattern '../'");
```
Comment (tavplubix) [inline src/DataTypes/TimezoneMixin.cpp:23]: It allows `..` and `../..`

**Review by tavplubix — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Let's also check the sources of CCTZ, maybe easier to forbid it there.

**Comment by Avogar** [comment]
Ok lets do it inside cctz library: https://github.com/ClickHouse/cctz/pull/7

**Pull request merged**